plutus-scb/src/Plutus/SCB/Webserver/WebSocket.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE MonoLocalBinds    #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# LANGUAGE TypeOperators     #-}
+
+module Plutus.SCB.Webserver.WebSocket
+    ( handleWS
+    ) where
+
+import           Control.Concurrent            (forkIO, threadDelay)
+import           Control.Exception             (SomeException, handle)
+import           Control.Monad                 (forever, void)
+import           Control.Monad.Freer           (Eff, LastMember, Member)
+import           Control.Monad.Freer.Extra.Log (Log, logInfo)
+import           Control.Monad.Freer.Reader    (Reader, ask)
+import           Control.Monad.Freer.WebSocket (WebSocketEffect, acceptConnection, receiveJSON, sendJSON)
+import           Control.Monad.IO.Class        (MonadIO, liftIO)
+import           Control.Monad.Logger          (LogLevel (LevelDebug))
+import qualified Data.Aeson                    as JSON
+import qualified Data.Aeson.Types              as JSON
+import qualified Data.Text                     as Text
+import           Data.Time.Units               (Second, TimeUnit, toMicroseconds)
+import           Network.WebSockets.Connection (Connection, PendingConnection, withPingThread)
+import           Plutus.SCB.App                (runApp)
+import           Plutus.SCB.Types              (Config, ContractExe)
+import           Plutus.SCB.Utils              (tshow)
+import           Plutus.SCB.Webserver.Handler  (getChainReport)
+import           Plutus.SCB.Webserver.Types    (StreamToClient (Echo, ErrorResponse, NewChainReport),
+                                                StreamToServer (Ping, Request))
+import           Wallet.Effects                (ChainIndexEffect)
+
+threadApp ::
+       ( Member ChainIndexEffect effs
+       , Member WebSocketEffect effs
+       , LastMember m effs
+       , MonadIO m
+       )
+    => Connection
+    -> Eff effs ()
+threadApp connection = do
+    chainReport <- getChainReport @ContractExe
+    sendJSON connection $ JSON.toJSON $ NewChainReport chainReport
+    threadDelayTime (10 :: Second)
+
+threadDelayTime ::
+       (LastMember m effs, MonadIO m, TimeUnit a) => a -> Eff effs ()
+threadDelayTime = liftIO . threadDelay . fromIntegral . toMicroseconds
+
+runThreadApp :: Config -> Connection -> IO ()
+runThreadApp env connection =
+    handle disconnect .
+    withPingThread connection 30 (pure ()) .
+    void . runApp LevelDebug env . forever $
+    threadApp connection
+  where
+    disconnect :: SomeException -> IO ()
+    disconnect _ = pure ()
+
+-- TODO Remove hard-coded `t ~ ContractExe`.
+handleClient ::
+       ( Member WebSocketEffect effs
+       , Member (Reader Config) effs
+       , LastMember m effs
+       , MonadIO m
+       )
+    => Connection
+    -> Eff effs ()
+handleClient connection = do
+    env <- ask
+    liftIO $ void $ forkIO $ void $ runThreadApp env connection
+    payload :: Either String JSON.Value <- receiveJSON connection
```

Comment:
I'm a bit confused by the fact that `connection` is used twice

1. In the forked thread inside the loop where we call `threadApp` every ten seconds, sending a `NewChainReport` without waiting for any messages for the client
2. In `handleClient` where we wait for a single message from the client, call `handleStreamingMessage` and then return the response on the connection

Does (2) not mess up the order of messages in (1) on the client side? (Also I don't know much about websockets!)

Other than that I think the way that `forkIO` is used with `runThreadApp` is fine. You could add an effect

```haskell
data Fork eff r where
  Fork :: Eff eff r -> Fork ThreadId
```

if you want to remove the `MonadIO` stuff.

Reply:
Working with the `eff` parameter can be quite annoying though.

Reply:
Actually maybe you could roll that into the `WebSocketEffect`:

```
data WebSocketEffect eff r where
  (...)
  AcceptConnection :: (Connection -> Eff eff ()) -> WebSocketEffect eff ()
```

and the handler for `AcceptConnection` could include the `handleDisconnect` and `withPingThread` calls from `runThreadApp`